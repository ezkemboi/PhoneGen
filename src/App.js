import React from 'react';
import './App.scss';

class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      numbersGenerated: [],
      amount: 0
    }
  }

  componentDidMount() {
    // Fetch stored phone numbers from local storage
    const contacts = JSON.parse(localStorage.getItem("contacts"));
    if (contacts && contacts.length >= 1) {
      this.setState({
        numbersGenerated: contacts
      });
    }
  }

  handleChange = (value) => {
    this.setState({
      amount: value
    })
  }

  // generatePdf = () => {
  //   console.log('I am clicked')
  //   const docItems = document.getElementsByName('contacts-table-list')
  //   window.onbeforeprint = (event) => {
  //     console.log('------>>>>>', event)
  //     document.write(docItems);
  //     document.close();
  //   }
  //   window.print()
  // }

  // Generate new phone number function
  generateNewPhoneNumber = () => {
    // Generate 9 random numbers
    let number;
    const { numbersGenerated, amount } = this.state;
    for (let i = 0; i < amount; i++) {
      number = `0${Math.floor(Math.random() * 1000000000)}`;
      // Check already created numbers and its length (MAX = 10,000)
      if (!numbersGenerated.includes(number) && numbersGenerated.length <= 10000) {
        numbersGenerated.push(number);
      }
    }
    // Add generated numbers to state and save local storage
    localStorage.setItem("contacts", JSON.stringify(numbersGenerated));
    this.setState({
      numbersGenerated
    });
  }

  render() {
    const { amount, numbersGenerated } = this.state;
    let phoneNumbers;
    let disabledButton;
    // Edge cases for catching when to disable button
    const itemsInLocalStorage = localStorage.getItem("contacts");
    if (amount < 1 || amount >= 10000 || (itemsInLocalStorage && itemsInLocalStorage.length >= 10000)) {
      disabledButton = "disabled"
    }
    if (numbersGenerated && numbersGenerated.length >= 1) {
      phoneNumbers = numbersGenerated.map((number, index) => {
        return (
          <div className="phone-list-table" key={index}>
            <div className="phone-list-table__left">
              <p>{index + 1}</p>
            </div>
            <div className="phone-list-table__right">
              <p>{number}</p>
            </div>
          </div>
        )
      });
    }

    return (
      <div>
        <h3 id="app-title">PhoneGen - Phone Generator App</h3>
        <div className="App">
          <div className="App__left">
            <h3>PhoneGen</h3>
            <p id="app-description">
              Welcome to new phone number generator app.
              Generate and store upto 10,000 phone numbers.
            </p>
            <input
              type="number"
              value={amount}
              onChange={e => this.handleChange(e.target.value)}
              name="phone"
            />
            <button
              id="generate-numbers"
              onClick={() => this.generateNewPhoneNumber()}
              disabled={disabledButton}
            >
              Generate
            </button>
          </div>
          <div className="App__right">
            <div className="download-pdf-file">
              {/* <button onClick={this.generatePdf}>Download As Pdf</button> */}
              <button>Download As Pdf</button>
            </div>
            <div className="headers-for-phones">
              <div className="headers-for-phones__index">
                <h3>Index</h3>
              </div>
              <div className="headers-for-phones__number">
                <h3>Phone Number</h3>
              </div>
            </div>
            <div className="contacts-table-list">
              {phoneNumbers ? phoneNumbers : <div>No contacts. Please generate contacts.</div>}
            </div>
          </div>
        </div>
      </div>
    )
  }
}

export default App;
