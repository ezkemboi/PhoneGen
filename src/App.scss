.App {
  display: flex;
  margin: 2% 10%;
  min-height: 85vh;
  max-height: 90vh;
  text-align: center;
  &__left {
    display: flex;
    flex-direction: column;
    width: 30%;
    margin-right: 30px;
    background-color: white;
    border: 1px solid gray;
    h3 {
      border-bottom: 1px solid currentColor;
      padding-bottom: 20px;
      width: 50%;
      align-self: center;
    }
    #app-description {
      margin: 20px;
      padding: 0 10px;
    }
    input {
      height: 30px;
      margin: 10px 20px;
      width: 40%;
      align-self: center;
      text-align: center;
    }
    button {
      display: flex;
      align-self: center;
      width: 70%;
      justify-content: center;
    }
    button:disabled:hover {
      background-color: white;
      color: black;
      opacity: 0.3;
    }
  }
  &__right {
    display: flex;
    background-color: white;
    width: 70%;
    flex-direction: column;
    border: 1px solid gray;
    .right-header-section {
      display: flex;
      justify-content: flex-end;
      margin: 10px;
      .clear-button {
        margin-right: 10px;
        border: 2px solid black;
        border-radius: 8px;
      }
      .clear-button:hover {
        background-color: red;
      }
      .order-contacts {
        display: flex;
        margin-right: 10px;
        align-self: center;
        font-size: 16px;
        border: 2px solid black;
        border-radius: 8px;
        height: 40px;
      }
    }
    .headers-for-phones {
      display: flex;
      border-bottom: 1px solid gray;
      border-top: 1px solid gray;
      &__index {
        width: 30%;
      }
      &__number {
        width: 70%;
      }
    }
    .contacts-table-list {
      overflow-y: scroll;
      .phone-list-table {
        display: flex;
        padding: 0 20px;
        &__left {
          width: 30%;
          p {
            border-bottom: 1px solid gray;
            padding-bottom: 2px;
          }
        }
        &__right {
          width: 70%;
          p {
            border-bottom: 1px solid gray;
            padding-bottom: 2px;
          }
        }
      }
    }
  }
  button {
    background-color: white;
    height: 40px;
    color: black;
    font-size: 16px;
  }
  button:hover {
    cursor: pointer;
    background: black;
    color: white;
  }
}

#app-title {
  text-align: center;
  font-size: 30px;
  line-height: 20px;
}
