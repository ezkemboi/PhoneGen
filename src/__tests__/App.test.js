import React from 'react';
import { shallow } from 'enzyme';
import App from '../App';

describe('APP', () => {
  const component = shallow(<App />);
  const contacts = [
    "024776563",
    "046455020",
    "0943165218"
  ]
  const event = { target: { value: 3 } }
  const option = { target: { value: 'Descending' } }
  const ascendingOption = { target: { value: 'Ascending' } }

  // Setting local storage
  jest.spyOn(window.localStorage.__proto__, 'setItem');
  jest.spyOn(window.localStorage.__proto__, 'getItem');
  jest.spyOn(window.localStorage.__proto__, 'clear');
  jest.useFakeTimers();

  // Set some number of contacts in local storage
  beforeEach(() => {
    localStorage.setItem("contacts", JSON.stringify(contacts));
  });

  it('renders without crashing including Component Did Mount Call', () => {
    const getComponent = () => shallow(<App />);
    // Call method above so that it will call componentDidMount()
    getComponent();
    expect(component).toMatchSnapshot();
    expect(component.length).toEqual(1)
  });

  it('generates contacts', () => {
    component.find('input#handle-change-input').simulate('change', event);
    component.find('button#generate-numbers').simulate('click');
    expect(localStorage.setItem).toHaveBeenCalled();
  });

  it('clears contacts from local storage', () => {
    component.find('button.clear-button').simulate('click');
    const contacts = JSON.parse(localStorage.getItem("contacts"));
    expect(localStorage.clear).toHaveBeenCalled();
    expect(contacts).toBeNull();
  });

  it('orders contacts in descending order', () => {
    component.find('select.order-contacts').simulate('change', option);
    const contacts = JSON.parse(localStorage.getItem("contacts"));
    expect(contacts).toBeInstanceOf(Array);
    expect(contacts.length).toBe(3);
    expect(localStorage.getItem).toHaveBeenCalled();
  });

  it('orders contacts in ascending order', () => {
    component.find('select.order-contacts').simulate('change', ascendingOption);
    const contacts = JSON.parse(localStorage.getItem("contacts"));
    expect(localStorage.getItem).toHaveBeenCalled();
    expect(contacts).toBeInstanceOf(Array);
    expect(contacts.length).toBe(3);
  });
});
