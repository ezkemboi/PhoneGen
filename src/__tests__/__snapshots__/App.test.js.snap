// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APP renders without crashing 1`] = `ShallowWrapper {}`;

exports[`APP renders without crashing including Component Did Mount Call 1`] = `ShallowWrapper {}`;
